### PR TITLE
Support TimestampType

### PR DESCRIPTION
**Pull request by elvaliuliuliu**
This PR exposes `Timestamp` class and supports `TimestampType` as below:
1. Support `DataFrame.Collect()` for `TimestampType`.
2. Support `CreateDataFrame` that takes in `Timestamp`.
3. Support UDF that takes in `Timestamp` and UDF that returns `Timestamp`.

### CODE REVIEW

**Comment by imback82** [comment]
@elvaliuliuliu Is this ready for review?

**Comment by elvaliuliuliu** [comment]
> @elvaliuliuliu Is this ready for review?

Yes, this is ready for review! Looks like request reviewers function is removed from github, so I will ping you next time when wip pr is ready for review. Thanks!

File: src/csharp/Microsoft.Spark/Sql/CustomPicklers.cs [hunk 1]
```diff
@@ -40,4 +40,15 @@ public void pickle(object o, Stream outs, Pickler currentPickler)
             currentPickler.save(((Date)o).GetInterval());
         }
     }
+
+    /// <summary>
+    /// Custom pickler for Timestamp objects.
+    /// </summary>
+    internal class TimestampPickler : IObjectPickler
+    {
+        public void pickle(object o, Stream outs, Pickler currentPickler)
```
Comment (Niharikadutta) [inline src/csharp/Microsoft.Spark/Sql/CustomPicklers.cs]: Do we want to rename `outs` to either `outStream` or just `out`? `outs` reads a little funny to me.

Comment (elvaliuliuliu) [reply src/csharp/Microsoft.Spark/Sql/CustomPicklers.cs]: Looking through some other examples, `outs` has been used to stand for `output stream`.

File: src/csharp/Microsoft.Spark.E2ETest/UdfTests/UdfSimpleTypesTests.cs [hunk 2]
```diff
@@ -21,10 +22,28 @@ public class UdfSimpleTypesTests
...
+                    null,
+                    new Date(2020, 1, 1),
+                    new Timestamp(2020, 1, 1, 0, 0, 0, 0)
+                }));
+            data.Add(new GenericRow(
+                new object[]
+                {
+                    30,
+                    new Date(2020, 1, 2),
+                    new Timestamp(2020, 1, 2, 15, 30, 30, 0)
```
Comment (imback82) [inline src/csharp/Microsoft.Spark.E2ETest/UdfTests/UdfSimpleTypesTests.cs]: Can we use non zero value for microseconds?

File: src/csharp/Microsoft.Spark.E2ETest/UdfTests/UdfSimpleTypesTests.cs [hunk 3]
```diff
@@ -74,9 +92,74 @@ public void TestUdfWithReturnAsDateType()
...
+        {
+            Func<Column, Column> udf = Udf<Timestamp, string>(time => time.ToString());
+
+            Row[] rows = _df.Select(udf(_df["time"])).Collect().ToArray();
+            Assert.Equal(2, rows.Length);
+
+            var expected = new string[]
+            {
+                "2020-01-01 00:00:00.000000",
+                "2020-01-02 15:30:30.000000"
```
Comment (imback82) [inline src/csharp/Microsoft.Spark.E2ETest/UdfTests/UdfSimpleTypesTests.cs]: I assume this is the same format that Scala/Python uses for toString?

File: src/csharp/Microsoft.Spark.E2ETest/UdfTests/UdfSimpleTypesTests.cs [hunk 4]
```diff
@@ -74,9 +92,74 @@ public void TestUdfWithReturnAsDateType()
...
+        }
+
+        /// <summary>
+        /// UDF that returns Timestamp type.
+        /// </summary>
+        [Fact]
+        public void TestUdfWithReturnAsTimestampType()
+        {
+            Func<Column, Column> udf1 = Udf<int?, Timestamp>(
+                s => new Timestamp(2020 + s.GetValueOrDefault(), 1, 4, 0, 0, 0, 0));
```
Comment (imback82) [inline src/csharp/Microsoft.Spark.E2ETest/UdfTests/UdfSimpleTypesTests.cs]: nit: Looks like this is carried from the existing code, but let's either use `x` or `i` (for integer) instead of `s`. (or even `age` is fine).

File: src/csharp/Microsoft.Spark/Sql/SparkSession.cs [hunk 5]
```diff
@@ -196,6 +196,15 @@ public void Dispose()
         public DataFrame CreateDataFrame(IEnumerable<Date> data) =>
             CreateDataFrame(ToGenericRows(data), SchemaWithSingleColumn(new DateType()));
 
+        /// <summary>
+        /// Creates a Dataframe given data as <see cref="IEnumerable"/> of type
+        /// <see cref="Timestamp"/>
+        /// </summary>
+        /// <param name="data"></param>
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/Sql/SparkSession.cs]: let's fill in the comment.

File: src/csharp/Microsoft.Spark/Sql/Types/SimpleTypes.cs [hunk 6]
```diff
@@ -80,10 +80,22 @@ internal override object FromInternal(object obj)
...
     {
+        private static readonly DateTime s_unixTimeEpoch =
+            new DateTime(1970, 1, 1, 0, 0, 0, DateTimeKind.Utc);
+
+        internal override bool NeedConversion() => true;
+
+        internal override object FromInternal(object obj)
+        {
+            return new Timestamp(new DateTime((long)obj * 10 + s_unixTimeEpoch.Ticks,
+                DateTimeKind.Utc));
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/Sql/Types/SimpleTypes.cs]: Can you add a short description of what value `obj`  holds as `long` and what this conversion does?

File: src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+{
+    /// <summary>
+    /// Represents Timestamp containing year, month, day, hour, minute, second, microsecond in
+    /// Coordinated Universal Time (UTC).
+    /// </summary>
+    public class Timestamp
+    {
+        private readonly DateTime _dateTime;
+        private static readonly DateTime s_unixTimeEpoch =
+            new DateTime(1970, 1, 1, 0, 0, 0, DateTimeKind.Utc);
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs]: Maybe, we should put this in a common place?

File: src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+
+        /// <summary>
+        /// Constructor for Timestamp class.
+        /// </summary>
+        /// <param name="dateTime">DateTime object</param>
+        public Timestamp(DateTime dateTime)
+        {
+            _dateTime = dateTime.ToUniversalTime();
+
+            if (dateTime.Kind != DateTimeKind.Utc)
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs]: You can check this before the above line. Can you also add a test for this?

File: src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+            {
+                throw new InvalidTimeZoneException(
+                    $"Invalid TimeZone for Timestamp, please use " +
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs]: Can you also print out the timezone set in `dateTime`?

File: src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+        /// Coordinated Universal Time (UTC).
+        /// </summary>
+        /// <param name="year">The year (1 through 9999)</param>
+        /// <param name="month">The month (1 through 12)</param>
+        /// <param name="day">The day (1 through the number of days in month)</param>
+        /// <param name="hour">The hour (0 through 23)</param>
+        /// <param name="minute">The minute (0 through 59)</param>
+        /// <param name="second">The second (0 through 59)</param>
+        /// <param name="microsecond">The microsecond (0 through 999999)</param>
+        public Timestamp(int year, int month, int day, int hour, int minute, int second, int microsecond)
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs]: over 100 limit?

File: src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+        {
+            if (microsecond <= 999999 && microsecond >= 0)
+            {
+                _dateTime = new DateTime(year, month, day, hour, minute, second)
+                    .AddTicks(microsecond * 10).ToUniversalTime();
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs]: Shouldn't the given year, month, etc. already be in UTC? Why do we need to call `ToUniversalTime`? I think you need to pass `DateTimeKind` when you constructor the `DateTime` object.

File: src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+            }
+            else
+            {
+                throw new ArgumentOutOfRangeException($"Invalid microsecond value {microsecond}");
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs]: Shall we put the valid range in the exception message as well?

File: src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+
+        /// <summary>
+        /// Returns the second component of the timestamp.
+        /// </summary>
+        public int Second => _dateTime.Second;
+
+        /// <summary>
+        /// Returns the microsecond component of the timestamp.
+        /// </summary>
+        public int Microsecond => (int)(_dateTime.Ticks -
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs]: Can we test this as well? Let's add a test case for `Timestamp` class.

File: src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+            new DateTime(Year, Month, Day, Hour, Minute, Second).Ticks) / 10;
+
+        /// <summary>
+        /// Readable string representation for this type.
+        /// </summary>
+        public override string ToString() => _dateTime.ToUniversalTime()
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs]: We should assume `_dateTime` should be already in UTC so that we don't need to call `ToUniversalTime`.

File: src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+        /// Checks if the given object is same as the current object.
+        /// </summary>
+        /// <param name="obj">Other object to compare against</param>
+        /// <returns>True if the other object is equal</returns>
+        public override bool Equals(object obj) =>
+            ReferenceEquals(this, obj) ||
+            ((obj is Timestamp timestamp) && Year.Equals(timestamp.Year) &&
+                Month.Equals(timestamp.Month) && Day.Equals(timestamp.Day) &&
+                Hour.Equals(timestamp.Hour) && Minute.Equals(timestamp.Minute) &&
+                Second.Equals(timestamp.Second) && Microsecond.Equals(timestamp.Microsecond));
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs]: Just compare `_dateTime`?

File: src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+
+        /// <summary>
+        /// Returns the hash code of the current object.
+        /// </summary>
+        /// <returns>The hash code of the current object</returns>
+        public override int GetHashCode() => base.GetHashCode();
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs]: Let's return a hash code of `_dateTime`.

File: src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+
+        /// <summary>
+        /// Returns DateTime object describing this type.
+        /// </summary>
+        /// <returns>DateTime object of the current object</returns>
+        public DateTime ToDateTime() => _dateTime;
+
+        /// <summary>
+        /// Returns a double object that represents the number of seconds from the epoch of
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs]: Can you convey the intention that the unit is seconds but we preserve the sub-seconds as fraction?

File: src/csharp/Microsoft.Spark.E2ETest/UdfTests/UdfSimpleTypesTests.cs (hunk 3 above)
Comment (elvaliuliuliu) [reply src/csharp/Microsoft.Spark.E2ETest/UdfTests/UdfSimpleTypesTests.cs]: Yes, same format.

File: src/csharp/Microsoft.Spark.E2ETest/UdfTests/UdfSimpleTypesTests.cs (hunk 2 above)
Comment (elvaliuliuliu) [reply src/csharp/Microsoft.Spark.E2ETest/UdfTests/UdfSimpleTypesTests.cs]: Got it, good catch! So actually non zero microseconds doesn't work for udf, it always converts to zero microseconds when reading input data [here](https://github.com/dotnet/spark/blob/9275678a6e46e06a266e5516b3fea50896ede761/src/csharp/Microsoft.Spark.Worker/Command/SqlCommandExecutor.cs#L135).

File: src/csharp/Microsoft.Spark.E2ETest/UdfTests/UdfSimpleTypesTests.cs (hunk 4 above)
Comment (elvaliuliuliu) [reply src/csharp/Microsoft.Spark.E2ETest/UdfTests/UdfSimpleTypesTests.cs]: Sure, changed for both here and previous code from `Date`!

File: src/csharp/Microsoft.Spark/Sql/SparkSession.cs (hunk 5 above)
Comment (elvaliuliuliu) [reply src/csharp/Microsoft.Spark/Sql/SparkSession.cs]: Sure, added!

File: src/csharp/Microsoft.Spark/Sql/Types/SimpleTypes.cs (hunk 6 above)
Comment (elvaliuliuliu) [reply src/csharp/Microsoft.Spark/Sql/Types/SimpleTypes.cs]: Sure, added.

File: src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs (hunk 7 above)
Comment (elvaliuliuliu) [reply src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs]: This is also used in [DateTimeExtension.cs](https://github.com/dotnet/spark/blob/9275678a6e46e06a266e5516b3fea50896ede761/src/csharp/Microsoft.Spark.Worker/Utils/DateTimeExtension.cs#L13), do you mean this?

File: src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs (hunk 9 above)
Comment (elvaliuliuliu) [reply src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs]: Added.

File: src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs (hunk 10 above)
Comment (elvaliuliuliu) [reply src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs]: Done.

File: src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs (hunk 11 above)
Comment (elvaliuliuliu) [reply src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs]: So pass `DateTimeKind` does not give the right time, still have 8 hours lack in this case so I have to convert it using `ToUniversalTime`.

File: src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs (hunk 12 above)
Comment (elvaliuliuliu) [reply src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs]: Sounds good! Added.

File: src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs (hunk 14 above)
Comment (elvaliuliuliu) [reply src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs]: This is the same as what happened above regarding passing kind is not working.

File: src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs (hunk 15 above)
Comment (elvaliuliuliu) [reply src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs]: Changed.

File: src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs (hunk 17 above)
Comment (elvaliuliuliu) [reply src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs]: Sure, I have changed the summary and let me know if this makes sense.

File: src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs (hunk 13 above)
Comment (elvaliuliuliu) [reply src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs]: Should I add a unit test for `Timestamp` class?

File: src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs (hunk 8 above)
Comment (elvaliuliuliu) [reply src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs]: Added this in Unit test.

File: src/csharp/Microsoft.Spark.UnitTest/Sql/TimestampTests.cs [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+namespace Microsoft.Spark.UnitTest
+{
+    public class TimestampTests
+    {
+        [Fact]
+        public void TimestampTest()
+        {
+            {
+                var timestamp = new Timestamp(
+                    new DateTime(2020, 1, 2, 15, 30, 30, DateTimeKind.Utc));
```
Comment (imback82) [inline src/csharp/Microsoft.Spark.UnitTest/Sql/TimestampTests.cs]: Can you add a test where `DataTime` has non zero microsecond (using `Ticks`)?

File: src/csharp/Microsoft.Spark.UnitTest/Sql/TimestampTests.cs [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+                Assert.Equal(15, timestamp.Hour);
+                Assert.Equal(30, timestamp.Minute);
+                Assert.Equal(30, timestamp.Second);
+                Assert.Equal(0, timestamp.Microsecond);
+
+                // Validate ToString().
+                Assert.Equal("2020-01-02 15:30:30.000000", timestamp.ToString());
+
+                // Validate ToDateTime().
+                Assert.IsType<DateTime>(timestamp.ToDateTime());
```
Comment (imback82) [inline src/csharp/Microsoft.Spark.UnitTest/Sql/TimestampTests.cs]: we need to verify the returned value is same is the one you started with.

Comment (imback82) [reply src/csharp/Microsoft.Spark.UnitTest/Sql/TimestampTests.cs]: I think we also need to verify that even if you pass in non-UTC `DateTime`, you get the UTC one.

File: src/csharp/Microsoft.Spark.UnitTest/Sql/TimestampTests.cs [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+            }
+
+            {
+                // Validate Microsecond and TimeZone.
+                var timestamp = new Timestamp(2020, 1, 2, 15, 30, 30, 123456);
+
+                Assert.Equal(123456, timestamp.Microsecond);
```
Comment (imback82) [inline src/csharp/Microsoft.Spark.UnitTest/Sql/TimestampTests.cs]: Can we just validate all fields? How about the invalid microsecond case?

File: src/csharp/Microsoft.Spark/Sql/Types/SimpleTypes.cs [hunk 21]
```diff
@@ -69,21 +69,52 @@ public sealed class BooleanType : AtomicType
     /// </summary>
     public sealed class DateType : AtomicType
     {
-        private static readonly DateTime s_unixTimeEpoch = new DateTime(1970, 1, 1);
+        internal static readonly DateTime s_unixTimeEpoch = new DateTime(1970, 1, 1);
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/Sql/Types/SimpleTypes.cs]: I think we can just set this to `new DateTime(1970, 1, 1, 0, 0, 0, DateTimeKind.Utc);` and use from `TimeStamp` as well.

File: src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs [hunk 22]
```diff
@@ -0,0 +1,140 @@
...
+            if (microsecond <= 999999 && microsecond >= 0)
+            {
+                // Create DateTime and AddTicks based on the microsecond value.
+                _dateTime = new DateTime(year, month, day, hour, minute, second, DateTimeKind.Utc)
+                    .AddTicks(microsecond * 10);
+            }
+            else
+            {
+                throw new ArgumentOutOfRangeException($"Invalid microsecond value {microsecond}. " +
+                    $"The microsecond is from 0 through 999999.");
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs]: nit: `The microsecond should be in the range of [0, 999999]`?

File: src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs [hunk 23]
```diff
@@ -0,0 +1,140 @@
...
+        public Timestamp(
+            int year,
+            int month,
+            int day,
+            int hour,
+            int minute,
+            int second,
+            int microsecond)
+        {
+            if (microsecond <= 999999 && microsecond >= 0)
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs]: nit: `if ((0 <= microsecond) && (microsecond <= 999999))`

File: src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs [hunk 24]
```diff
@@ -0,0 +1,140 @@
...
+
+        /// <summary>
+        /// Returns the second component of the timestamp.
+        /// </summary>
+        public int Second => _dateTime.Second;
+
+        /// <summary>
+        /// Returns the microsecond component of the timestamp.
+        /// </summary>
+        public int Microsecond => (int)(_dateTime.Ticks -
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs]: Can't you just do `(int)(_dateTime.Ticks % 10000000 / 10)`?

File: src/csharp/Microsoft.Spark/Sql/Types/SimpleTypes.cs (hunk 21 above)
Comment (elvaliuliuliu) [reply src/csharp/Microsoft.Spark/Sql/Types/SimpleTypes.cs]: Sure, I have changed all to use `DateType.s_unixTimeEpoch`.

File: src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs (hunk 22 above)
Comment (elvaliuliuliu) [reply src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs]: Changed.

File: src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs (hunk 23 above)
Comment (elvaliuliuliu) [reply src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs]: Changed!

File: src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs (hunk 24 above)
Comment (elvaliuliuliu) [reply src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs]: If we use `(int)(_dateTime.Ticks % 10000000 / 10)`, it will be the whole value represented in microseconds instead of just microsecond value, no?

Comment (imback82) [reply src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs]: Since we are modding, it should just get the microsecond value.

Comment (elvaliuliuliu) [reply src/csharp/Microsoft.Spark/Sql/Types/Timestamp.cs]: Yep, right! Thanks.

File: src/csharp/Microsoft.Spark.UnitTest/Sql/TimestampTests.cs (hunk 18 above)
Comment (elvaliuliuliu) [reply src/csharp/Microsoft.Spark.UnitTest/Sql/TimestampTests.cs]: Added!

File: src/csharp/Microsoft.Spark.UnitTest/Sql/TimestampTests.cs (hunk 19 above)
Comment (elvaliuliuliu) [reply src/csharp/Microsoft.Spark.UnitTest/Sql/TimestampTests.cs]: Added the tests, thanks!

File: src/csharp/Microsoft.Spark.UnitTest/Sql/TimestampTests.cs (hunk 20 above)
Comment (elvaliuliuliu) [reply src/csharp/Microsoft.Spark.UnitTest/Sql/TimestampTests.cs]: Yes, added for both!

File: src/csharp/Microsoft.Spark.UnitTest/Sql/TimestampTests.cs [hunk 25]
```diff
@@ -0,0 +1,76 @@
...
+namespace Microsoft.Spark.UnitTest
+{
+    public class TimestampTests
+    {
+        [Fact]
+        public void TimestampTest()
+        {
+            {
+                var timestamp = new Timestamp(
+                    new DateTime(2020, 1, 1, 8, 30, 30, DateTimeKind.Utc).AddTicks(1230));
```
Comment (imback82) [inline src/csharp/Microsoft.Spark.UnitTest/Sql/TimestampTests.cs]: Shall we create `var testDate = new DateTime(2020, 1, 1, 8, 30, 30, DateTimeKind.Utc).AddTicks(1230);` and reuse it below?

Comment (elvaliuliuliu) [reply src/csharp/Microsoft.Spark.UnitTest/Sql/TimestampTests.cs]: Sounds good. Changed!
